src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -80,18 +78,188 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
 
             var semanticVersion = await GetDocumentSemanticVersionAsync(documentSnapshot).ConfigureAwait(false);
 
-            // If we have a matching cached response, avoid computation and return early.
-            if (_cachedResponses.TryGetValue(textDocumentIdentifier.Uri, out var response) &&
-                response.IsCSharpFinalized &&
-                response.SemanticVersion == semanticVersion &&
-                response.Range == range)
+            // See if we can use our cache to at least partially avoid recomputation.
+            if (!_tokensCache.TryGetCachedTokens(textDocumentIdentifier.Uri, semanticVersion, range, out var cachedResult))
```

Comment:
Hmmm if we changed the API to be `GetCachedTokens` and it always returned a range which could possibly be 0 length would that enable this path to have a single token resolution mechanism?

Reply:
Hmm, I think we could but imo it would make things confusing. We would need to return a 0-length range, and the 0-length range's start/end would need to fall within the requested range. This seems like a too-specific implementation detail of the cache that I don't know whether our semantic tokens service should take a dependency on ☹️

Reply:
Ah I see, fair enough!